lib/connection/apm.js

```diff
@@ -0,0 +1,221 @@
+'use strict';
+const KillCursor = require('../connection/commands').KillCursor;
+const GetMore = require('../connection/commands').GetMore;
+
+/** Commands that we want to redact because of the sensitive nature of their contents */
+const SENSITIVE_COMMANDS = [
+  'authenticate',
+  'saslStart',
+  'saslContinue',
+  'getnonce',
+  'createUser',
+  'updateUser',
+  'copydbgetnonce',
+  'copydbsaslstart',
+  'copydb'
+];
+
+// helper methods
+const extractCommandName = command => Object.keys(command)[0];
+const calculateDuration = started => Date.now() - started;
+const generateConnectionId = pool => `${pool.options.host}:${pool.options.port}`;
+const maybeRedact = (commandName, result) =>
+  SENSITIVE_COMMANDS.indexOf(commandName) !== -1 ? {} : result;
+
+const LEGACY_FIND_QUERY_MAP = {
+  $query: 'filter',
+  $orderby: 'sort',
+  $hint: 'hint',
+  $comment: 'comment',
+  $maxScan: 'maxScan',
+  $max: 'max',
+  $min: 'min',
+  $returnKey: 'returnKey',
+  $showDiskLoc: 'showRecordId',
+  $maxTimeMS: 'maxTimeMS',
+  $snapshot: 'snapshot'
+};
+
+const LEGACY_FIND_OPTIONS_MAP = {
+  numberToSkip: 'skip',
+  numberToReturn: 'batchSize',
+  returnFieldsSelector: 'projection'
+};
+
+const OP_QUERY_KEYS = [
+  'tailable',
+  'oplogReplay',
+  'noCursorTimeout',
+  'awaitData',
+  'partial',
+  'exhaust'
+];
+
+/**
+ * Extract the actual command from the query, possibly upconverting if it's a legacy
+ * format
+ *
+ * @param {Object} command the command
+ */
+const extractCommand = command => {
+  if (command instanceof GetMore) {
+    return {
+      getMore: command.cursorId,
+      collection: command.ns.split('.')[1],
+      batchSize: command.numberToReturn
+    };
+  }
+
+  if (command instanceof KillCursor) {
+    return {
+      killCursors: command.ns.split('.')[1],
+      cursors: command.cursorIds
+    };
+  }
+
+  if (command.query && typeof command.query.$query !== 'undefined') {
+    // upconvert legacy find command
+    const result = { find: command.ns.split('.')[1] };
+    Object.keys(LEGACY_FIND_QUERY_MAP).forEach(key => {
+      if (typeof command.query[key] !== 'undefined')
+        result[LEGACY_FIND_QUERY_MAP[key]] = command.query[key];
+    });
+
+    Object.keys(LEGACY_FIND_OPTIONS_MAP).forEach(key => {
+      if (typeof command.options[key] !== 'undefined')
+        result[LEGACY_FIND_OPTIONS_MAP[key]] = command.options[key];
+    });
+
+    OP_QUERY_KEYS.forEach(key => {
+      if (command[key]) result[key] = command[key];
+    });
+
+    if (typeof command.pre32Limit !== 'undefined') {
+      result.limit = command.pre32Limit;
+    }
+
+    if (command.query.$explain) {
+      return { explain: result };
+    }
+
+    return result;
+  }
+
+  return command.query ? command.query : command;
+};
+
+const extractReply = (command, reply) => {
+  if (command instanceof GetMore) {
```

Comment:
How many functions do we have of the form:

```js
const myFunction = (command, ...) => {
  if (command instanceof GetMore) {
    return ...;
  }
  if (command instanceof KillCursor) {
    return ...;
  }
  if (command.query && typeof command.query.$query !== 'undefined') {
    return ...;
  }
  return ...;
};
```

Is this a pattern that we'd want to codify, or would that be over-optimizing?

Something like:

```js
const GET_MORE_COMMAND = Symbol();
const KILL_CURSOR_COMMAND = Symbol();
const FIND_COMMAND = Symbol();
const OTHER_COMMAND = Symbol();
const typeOfCommand = command => {
  if (command instanceof GetMore) {
    return GET_MORE_COMMAND;
  }
  if (command instanceof KillCursor) {
    return KILL_CURSOR_COMMAND;
  }
  if (command.query && typeof command.query.$query !== 'undefined') {
    return FIND_COMMAND;
  }
  return OTHER_COMMAND;
};

const extractCommand = command  => {
  switch (typeOfCommand(command)) {
    case GET_MORE_COMMAND: return ...;
    case KILL_CURSOR_COMMAND: return ...;
    case FIND_COMMAND: return ...;
    default: return ...;
  }
}
```

Reply:
we only have them here, and only as long as we support the old op codes (which are in the process of dying). I don't think there's compelling reason to further refactor this at this time.